ADR14  M1 update

chcmedeiros opened the pull request:
Support Milestone 1 of ADR14 implementation:
- Rename old methods;
- Add new methods.

ftheirs commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@oasisprotocol/ledger",
-  "version": "0.2.3",
+  "version": "0.2.4",
```
To keep consistency with the new methods, **getAddressAndPubKey** was renamed to **getAddressAndPubKey_ed25519** (new method is called **getAddressAndPubKey_secp256k1**)
To respect SemVer, major field should be bumped. However, if you don't want to make breaking changes in the API, it's possible to keep the old names.
I think that this should be decided by Oasis team 😄

chcmedeiros replied:
Good point! I will bump the major so that we have a version that respects SemVer, since the changes in this PR ask for that. Then Oasis Team can decide to merge this as is or we change the old method names.

pro-wh commented on src/helperV1.js:
```diff
@@ -29,7 +29,7 @@ export async function signSendChunkv1(app, chunkIdx, chunkNum, chunk) {
   }
   return app.transport
-    .send(CLA, INS.SIGN_ED25519, payloadType, 0, chunk, [0x9000, 0x6984, 0x6a80])
+    .send(CLA, ins, payloadType, 0, chunk, [0x9000, 0x6984, 0x6a80])
```
is this meant to be a new parameter like in the changes below?

chcmedeiros replied:
Thanks for noticing @pro-wh ! It is fixed now!

matevz commented:
I'm getting bunch of linter errors when running `yarn serve`. Can you check it out?

matevz commented:
> I'm getting bunch of linter errors when running `yarn serve`. Can you check it out?

Nevermind, it seems I had some leftovers from the old repo.

matevz commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@oasisprotocol/ledger",
-  "version": "0.2.3",
+  "version": "1.0.0",
```
v1.0.0 or v0.2.4? I would stick to the latter.

matevz reviewed (changes requested):
- please add proper secp256k1 tests (the pubkey and address should match the ones from Ethereum)
- please add the secp256k1 address and ParaTime transaction signing also to the UI (e.g. by adding 4 new buttons)
- Please fix the integration tests. I'm having issues running `yarn test:integration`. This https://github.com/chcmedeiros/ledger-js/blob/zondax-M1-update/tests/basic.ispec.js#L59 does not seem to be a vaild native Oasis address (it's too long). Should integration tests be working?

chcmedeiros commented:
@matevz I have fixed the old tests that were not correct and added tests coverage for the new methods. Also tested it using a NanoS with the test mnemonic).
Regarding the UI (.vue) that example has some time, and we have deprecated it months ago.

matevz commented on tests/basic.ispec.js:
```diff
@@ -4,176 +4,297 @@ import { createHash } from "crypto";
 
-  expect(resp.return_code).toEqual(0x9000);
-  expect(resp.error_message).toEqual("No errors");
+    expect(resp.hex_address).toEqual("7dac3ca6b185b7778a7027a52c0cc3940d2e1ca0");
```
The standard BIP44 address for testing mnemonic `equip will roof matter pink blind book anxiety banner elbow sun young` is `95e5e3C1BDD92cd4A0c14c62480DB5867946281D`. Where does discrepancy come from?

matevz commented on tests/basic.ispec.js:
```diff
@@ -4,176 +4,297 @@ import { createHash } from "crypto";
-  const transport = await TransportNodeHid.create(1000);
-  const app = new OasisApp(transport);
+    // Derivation path. First 3 items are automatically hardened!
+    const path = [44, 60, 0];
```
Please also cover the case for the 2nd key `[44', 60', 0', 0, 1]`. The helper function in `helperV1.js` will probably need to be updated or a new one added, but please don't break backward compatibility.

matevz commented:
I'm getting compilation error after applying https://github.com/oasisprotocol/ledger-js/pull/627/commits/ac681ad5b61bfd5d8b20cfabaa1d8c4dffb40d30. Can you take a look?

```sh
$ yarn build
yarn run v1.22.5
warning ../package.json: No license field
$ babel src --out-dir dist && yarn build-types
Successfully compiled 3 files with Babel (610ms).
warning ../package.json: No license field
$ tsc -p ./tsconfig.json
src/types.ts:39:30 - error TS2345: Argument of type 'Buffer | { return_code: number; error_message: string; } | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }' is not assignable to parameter of type 'Response<Buffer | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }>'.
  Type 'Buffer' is not assignable to type 'Response<Buffer | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }>'.
    Type 'Buffer' is not assignable to type 'Buffer & { return_code: number; error_message: string; }'.
      Type 'Buffer' is missing the following properties from type '{ return_code: number; error_message: string; }': return_code, error_message

39   console.log(successOrThrow(await app.getAddressAndPubKey_secp256k1([44])).hex_address.trim());
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/types.ts:39:77 - error TS2339: Property 'hex_address' does not exist on type 'Buffer | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }'.
  Property 'hex_address' does not exist on type 'Buffer'.

39   console.log(successOrThrow(await app.getAddressAndPubKey_secp256k1([44])).hex_address.trim());
                                                                               ~~~~~~~~~~~

src/types.ts:40:30 - error TS2345: Argument of type 'Buffer | { return_code: number; error_message: string; } | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }' is not assignable to parameter of type 'Response<Buffer | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }>'.

40   console.log(successOrThrow(await app.showAddressAndPubKey_secp256k1([44])).hex_address.trim());
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

src/types.ts:40:78 - error TS2339: Property 'hex_address' does not exist on type 'Buffer | { pk: Buffer; hex_address: string; return_code: any; error_message: string; }'.
  Property 'hex_address' does not exist on type 'Buffer'.

40   console.log(successOrThrow(await app.showAddressAndPubKey_secp256k1([44])).hex_address.trim());
                                                                                ~~~~~~~~~~~


Found 4 errors in the same file, starting at: src/types.ts:39
```

matevz reviewed (approved):
Great work, thanks.

pro-wh commented:
@matevz are we waiting for anything before merging this?

kostko commented:
Looks like the tests did not run for some reason.

matevz commented on src/index.js:
```diff
@@ -42,6 +42,21 @@ function processGetAddrEd25519Response(response) {
+  const returnCode = errorCodeData[0] * 256 + errorCodeData[1];
+
+  const pk = Buffer.from(response.slice(0, 33));
+  const hex_address = Buffer.from(response.slice(33, 73)).toString();
```
```suggestion
  const hexAddress = Buffer.from(response.slice(33, 73)).toString();
```

matevz commented on src/index.js:
```diff
@@ -42,6 +42,21 @@ function processGetAddrEd25519Response(response) {
+
+  return {
+    pk,
+    hex_address,
```
```suggestion
    hex_address: hexAddress,
```

matevz reviewed (changes requested):
I'm getting an error when running `yarn build-example`.  @chcmedeiros Can you take a look and then merge this so we can add ParaTime support for Ledger to our javascript wallets as well.

chcmedeiros commented:
> I'm getting an error when running `yarn build-example`. @chcmedeiros Can you take a look and then merge this so we can add ParaTime support for Ledger to our javascript wallets as well.

@matevz Regarding the vue-example that example has some time, and we have deprecated it months ago. If it's ok with you and since we have test coverage for the package with the real device we would remove that vue-example folder, avoiding more errors.

pro-wh commented:
sure, sounds good. would reduce the dev dependency tree by a lot too

chcmedeiros commented:
Vue example was removed. @matevz @pro-wh If you want to take a last look at it

chcmedeiros commented:
Hey @matevz The M3-maintenance branch now has a version that is working here on the ledger-js, if you want to test. Regarding the ledger tests, I think the procedure would be to compile the app from the repository to get the .elf and then use our tool zemu (device emulator) combined with this repo to run the tests. This should require some more work, so maybe we can merge this, discuss this tests implementation and possibly do it in the future PR for SR25519.

matevz commented:
> Hey @matevz The M3-maintenance branch now has a version that is working here on the ledger-js, if you want to test. Regarding the ledger tests, I think the procedure would be to compile the app from the repository to get the .elf and then use our tool zemu (device emulator) combined with this repo to run the tests. This should require some more work, so maybe we can merge this, discuss these tests implementation and possibly do it in the future PR for SR25519.

I can confirm that tests now work with 9b9c85a304c114222601104309ac3d38b13a1d12 . Feel free to merge this PR.

chcmedeiros commented:
@matevz I am not able to merge, probably don't haver permission to do so.
